### PR TITLE
Testing Implemented

### PR DESCRIPTION
**Pull request by louisbuchbinder**
Hey @jakech,

I have implemented the testing frameworks with our project. Check out the grunt file. `grunt mocha` will run the client side tests. `grunt mochaTest` will run the server side tests. `grunt watch will watch all js files and run all the tests`

Also I moved the algorithm stuff to app/chrome/

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,6 +26,7 @@
   "dependencies": {
     "body-parser": "^1.15.0",
     "express": "^4.13.4",
+    "grunt-mocha": "^1.0.0",
```
Comment (jakech) [inline package.json]: this should goto `devDependencies`

File: Gruntfile.js [hunk 2]
```diff
@@ -3,12 +3,28 @@ module.exports = function(grunt) {
...
+        },
+        src: [
+          'app/server/spec/**/*.js',
+          'app/client/spec/**/*.js',
+          'app/chrome/spec/**/*.js',
+        ]
+      }
+    },
+
+    mocha: {
```
Comment (jakech) [inline Gruntfile.js:20]: Lets keep all test in the command line. we dont need the run client tests on html

File: app/chrome/algorithm/spec/LinkedListSpec.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+describe('Example Tests', function () {
+  // it('should equal 2 in just plain mocha', function () {
+  //   assert.equal(2, 1 + 1);
+  // });
+  it('An example of using chai should', function () {
+    chai.should();
+    foo.should.be.a('string');
+    foo.should.equal('bar');
+    foo.should.have.length(3);
+    tea.should.have.property('flavors')
```
Comment (jakech) [inline app/chrome/algorithm/spec/LinkedListSpec.js]: missing `;`

File: app/chrome/algorithm/spec/LinkedListSpec.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    expect(foo).to.have.length(3);
+    expect(tea).to.have.property('flavors')
+      .with.length(3);
+  });
+  it('An example of using chai assert', function () {
+    var assert = chai.assert;
+
+    assert.typeOf(foo, 'string');
+    assert.equal(foo, 'bar');
+    assert.lengthOf(foo, 3)
```
Comment (jakech) [inline app/chrome/algorithm/spec/LinkedListSpec.js]: missing `;`

File: app/chrome/algorithm/spec/LinkedListSpec.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    tea.should.have.property('flavors')
+      .with.length(3);
+  });
+  it('An example of using chai expect', function () {
+    var expect = chai.expect;
+
+    expect(foo).to.be.a('string');
+    expect(foo).to.equal('bar');
+    expect(foo).to.have.length(3);
+    expect(tea).to.have.property('flavors')
```
Comment (jakech) [inline app/chrome/algorithm/spec/LinkedListSpec.js]: missing `;`

File: app/chrome/algorithm/spec/LinkedListSpec.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    assert.property(tea, 'flavors');
+    assert.lengthOf(tea.flavors, 3);
+  });
+});
+
+
+
+
+var assert = chai.assert;
+chai.should();
```
Comment (jakech) [inline app/chrome/algorithm/spec/LinkedListSpec.js:3]: a bit messy, would use a bit clean up

File: app/server/spec/test.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  it('An example of using chai assert', function () {
+		var assert = chai.assert;
+
+		assert.typeOf(foo, 'string');
+		assert.equal(foo, 'bar');
+		assert.lengthOf(foo, 3)
+		assert.property(tea, 'flavors');
+		assert.lengthOf(tea.flavors, 3);
+  });
+});
```
Comment (jakech) [inline app/server/spec/test.js:39]: [Best practice](http://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-file) to leave an empty line for every `.js` file.

**Comment by jakech** [comment]
@louisbuchbinder 
Good, but it's unnecessary to run test on html, we should have all tests run on command line.
Also, a bit of clean up would be great, use `jshint`.

**Comment by jakech** [comment]
http://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-file

**Comment by louisbuchbinder** [comment]
Thanks. The `grunt mocha` will run the client tests in the command line. The html file needs to be included because the `grunt mochaTest` uses  require('myFile.js'); like in node, but the client tests have to use <script src="myFile.js"></script>.

**Comment by jakech** [comment]
@louisbuchbinder 
you could still use `require()` inside your test scripts, as tests are basically run locally by node.

in the case of testing `LinkedList.js`

inside `/app/chrome/spec/LinkedListSpec.js`

``` javascript
require('../algorithm/LinkedList.js');

// test below
describe('Blah blah blah', function(){
  // tests
  it('blah', function(){
    var l = new LinkedList();
    l.shoud ...
  });
});
```

**Comment by louisbuchbinder** [comment]
Okay I see. so we can just do all the tests using the `grunt mochaTest`. I will change that.

I have cleaned up using the jshint, but the watch does not seem to be automatically running the js hint

**Comment by louisbuchbinder** [comment]
I am getting issues using the `require('../LinkedList.js')`. I believe this is due to the fact that LinkedList.js does not have any `module.exports`. Please advise...

**Comment by jakech** [comment]
@louisbuchbinder 
About the test, I think you are right, my bad. we need to use the `html`.
But I think we don't need a grunt task for that.

Or the grunt task would be just opening the `SpecRunner.html`

**Comment by louisbuchbinder** [comment]
no the grunt task does not actually open the html. It just runs it in the command line. Try running `grunt mocha` to see what I mean. Alternatively, you could open the html to view the test in a browser, but that way is not required for the test to run.

**Comment by jakech** [comment]
I see, that's cool. i will pull it down and try it right now.

**Comment by jakech** [comment]
ok, I got how it works now, thanks! I m merging it.

**Pull request merged**